File: lib/gotenberg.ts
function getGotenbergEndpoint(pathname: string): string {
  const baseUrl = process.env.GOTENBERG_URL

  if (!baseUrl) {
    throw new Error("GOTENBERG_URL is not configured")
  }

  return `${baseUrl.replace(/\/$/, "")}${pathname}`
}

function getAuthorizationHeader(): string | null {
  const username = process.env.GOTENBERG_USERNAME
  const password = process.env.GOTENBERG_PASSWORD

  if (!username || !password) {
    return null
  }

  return `Basic ${Buffer.from(`${username}:${password}`).toString("base64")}`
}

type GotenbergError = Error & {
  status?: number
  targetUrl?: string
}

export async function convertUrlToPdf(targetUrl: string): Promise<Buffer> {
  const formData = new FormData()
  formData.set("url", targetUrl)
  formData.set("printBackground", "true")
  formData.set("preferCssPageSize", "true")
  formData.set("waitForSelector", "[data-cv-print-ready='true']")

  const headers = new Headers()
  const authorization = getAuthorizationHeader()
  if (authorization) {
    headers.set("Authorization", authorization)
  }

  const response = await fetch(getGotenbergEndpoint("/forms/chromium/convert/url"), {
    method: "POST",
    headers,
    body: formData,
  })

  if (!response.ok) {
    const errorBody = await response.text().catch(() => "")
    const error = new Error(`Gotenberg request failed for ${targetUrl} with status ${response.status}: ${errorBody}`) as GotenbergError
    error.status = response.status
    error.targetUrl = targetUrl
    throw error
  }

  const pdfBytes = await response.arrayBuffer()
  return Buffer.from(pdfBytes)
}

File: backend/internal/platform/gotenberg/gotenberg.go
package gotenberg

import (
	"bytes"
	"encoding/base64"
	"fmt"
	"io"
	"mime/multipart"
	"net/http"
	"strings"
	"time"

	"github.com/hyunbridge/website/backend/internal/config"
)

var gotenbergHTTPClient = &http.Client{Timeout: 60 * time.Second}

func ConvertURLToPDF(cfg config.Config, targetURL string) ([]byte, error) {
	if cfg.GotenbergURL == "" {
		return nil, fmt.Errorf("gotenberg url is not configured")
	}

	var body bytes.Buffer
	writer := multipart.NewWriter(&body)
	_ = writer.WriteField("url", targetURL)
	_ = writer.WriteField("printBackground", "true")
	_ = writer.WriteField("preferCssPageSize", "true")
	_ = writer.WriteField("waitForSelector", "[data-cv-print-ready='true']")
	if err := writer.Close(); err != nil {
		return nil, err
	}

	endpoint := strings.TrimSuffix(cfg.GotenbergURL, "/") + "/forms/chromium/convert/url"
	req, err := http.NewRequest(http.MethodPost, endpoint, &body)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", writer.FormDataContentType())
	if cfg.GotenbergUsername != "" && cfg.GotenbergPassword != "" {
		creds := base64.StdEncoding.EncodeToString([]byte(cfg.GotenbergUsername + ":" + cfg.GotenbergPassword))
		req.Header.Set("Authorization", "Basic "+creds)
	}

	resp, err := gotenbergHTTPClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		payload, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		return nil, fmt.Errorf("gotenberg request failed with status %d: %s", resp.StatusCode, string(payload))
	}

	return io.ReadAll(resp.Body)
}
